perception/image_projection_based_fusion/src/fusion_node.cpp

```diff
@@ -157,67 +143,157 @@ void FusionNode<Msg, Obj>::preprocess(Msg & ouput_msg __attribute__((unused)))
+            "debug/timestamp_intervel_ms/roi" + std::to_string(roi_i), timestamp_intervel_ms);
+          debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+            "debug/timestamp_intervel_offset_ms/roi" + std::to_string(roi_i),
+            timestamp_intervel_ms - input_offset_ms_.at(roi_i));
+        }
+      }
     }
+  }
+
+  // if all camera fused, postprocess; else, cache the Msg
+  if (std::count(is_fused_.begin(), is_fused_.end(), true) == int(rois_number_)) {
+    timer_->cancel();
+    postprocess(*output_msg);
+    publish(*output_msg);
+    std::fill(is_fused_.begin(), is_fused_.end(), false);
 
-    fuseOnSingleImage(
-      *input_msg, image_id, *input_roi_msg, camera_info_map_.at(image_id), output_msg);
+    // add processing time for debug
+    if (debug_publisher_) {
+      const double cyclic_time_ms = stop_watch_ptr_->toc("cyclic_time", true);
+      const double processing_time_ms = stop_watch_ptr_->toc("processing_time", true);
+      debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+        "debug/cyclic_time_ms", cyclic_time_ms);
+      debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+        "debug/processing_time_ms", processing_time_ms);
+    }
+  } else {
+    sub_stdpair_.first = int64_t(timestamp_nsec);
+    sub_stdpair_.second = output_msg;
   }
+}
+
+template <class Msg, class Obj>
+void FusionNode<Msg, Obj>::roiCallback(
+  const DetectedObjectsWithFeature::ConstSharedPtr input_roi_msg, const std::size_t roi_i)
+{
+  int64_t timestamp_nsec =
+    (*input_roi_msg).header.stamp.sec * (int64_t)1e9 + (*input_roi_msg).header.stamp.nanosec;
+
+  // if cached Msg exist, try to match
+  if (sub_stdpair_.second != nullptr) {
+    int64_t newstamp = sub_stdpair_.first + input_offset_ms_.at(roi_i) * (int64_t)1e6;
+    int64_t interval = abs(timestamp_nsec - newstamp);
+
+    if (interval < match_threshold_ms_ * (int64_t)1e6 && is_fused_.at(roi_i) == false) {
+      if (camera_info_map_.find(roi_i) == camera_info_map_.end()) {
+        RCLCPP_WARN(this->get_logger(), "no camera info. id is %zu", roi_i);
+        (roi_stdmap_.at(roi_i))[timestamp_nsec] = input_roi_msg;
+        return;
+      }
+      if (debugger_) {
+        debugger_->clear();
+      }
+
+      fuseOnSingleImage(
+        *(sub_stdpair_.second), roi_i, *input_roi_msg, camera_info_map_.at(roi_i),
+        *(sub_stdpair_.second));
+      is_fused_.at(roi_i) = true;
+
+      if (debug_publisher_) {
+        double timestamp_intervel_ms = (timestamp_nsec - sub_stdpair_.first) / 1e6;
+        debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+          "debug/timestamp_intervel_ms/roi" + std::to_string(roi_i), timestamp_intervel_ms);
+        debug_publisher_->publish<tier4_debug_msgs::msg::Float64Stamped>(
+          "debug/timestamp_intervel_offset_ms/roi" + std::to_string(roi_i),
+          timestamp_intervel_ms - input_offset_ms_.at(roi_i));
```

Comment:
[imo] The output topic name should be like `debug/roi0/timestamp_intervel_ms` since the last field shows the topic type.

Reply:
fixed in https://github.com/autowarefoundation/autoware.universe/pull/2175/commits/7807a448f3c772d49ebe50bc6f5fe3863ead6a64